### DISCUSSION TITLE
eCAL C API Usage and Documentation

### DISCUSSION THREAD

**Discussion by chengguizi** [eclipse-ecal/ecal#796] [General]
May I quickly check if the eCAL C API is documented anywere, and is it consider stable for use? I am trying to see if eCAL can be compiled into embedded chip like STM32 which runs on RTOS.

I know techinically we can use C++ to programm STM32, for some RTOS too (e.g. Nuttx). But just wondering, for eCAL team internally, have you guys put eCAL on to STM32 already, and which RTOS to recommend trying?

**Comment by FlorianReimold** [comment]
eCAL compiles (and runs) on pretty much anything that is POSIX (-like) and on Microsoft Windows. You may encounter some difficulties though, but maybe it's worth a try. However I don't think however that we have ever run eCAL on that Processor. We have binaries for Ubuntu armhf, so it does compile on a 32bit ARM architecture.
The C API is stable, yes. If we change anything on that, we always aim to make it backwards compatible and indiciate the changes with an increase of the Minor version.

The Documentation is unfortunately mixed with the C++ API Documentation. But here you should find most of what you need:
https://eclipse-ecal.github.io/ecal/_api/file_ecal_core_include_ecal_ecal_clang.h.html

**Comment by chengguizi** [comment]
Thanks @FlorianReimold ,

Sounds like C API should be tried out first! Is it considered full-feature compared to the C++ counterpart?

**Comment by KerstinKeller** [comment]
To be clear here, the C API is realized as a wrapper around the C++ API. In the past we used it for creating language bindings to other languages (e.g. Python) and when we needed full ABI compatibility between different compilers on the same OS (e.g. between VS 2012 / 2015 applications).

However, it is not a pure C implementation of eCAL.

We don't have experience running eCAL on RTOS. Whether you will be able to run eCAL there will depend highly on the resources of the compiler, both RAM and ROM.
The pure communation core only has few dependencies, the largest one being protobuf. Protobuf on its own is very resource hungry.

There used to be a stripped down version (ecal_metal) to adress this, however we have removed it again (for various reasons).

In general, you might be better of using a protocol like MQTT for connection to RTOS devices, and then bridge the traffic to eCAL. (see also https://github.com/eclipse-ecal/ecal-mqtt-bridge).

**Comment by chengguizi** [comment]
Hi @KerstinKeller ,

That makes a lot of sense! Yes, I shouldn't have expected the eCAL will run directly on embedded hardware which don't normally come with network protocol (TCP/IP). And noted on the power hungryness of the serialisation protocol such as protobuf.

I shall stay using eCAL only for Linux-capable machines then, and use other dedicated embedded protocol for STM32-based communication.
